src/client/datascience/raw-kernel/rawJupyterSession.ts

```diff
@@ -47,15 +63,86 @@ export class RawJupyterSession extends BaseJupyterSession {
         throw new Error('Not implemented');
     }
 
-    // RAWKERNEL: Cancel token routed down?
-    public async connect(connectionInfo: IJMPConnectionInfo, _cancelToken?: CancellationToken): Promise<void> {
-        await this.rawSession.connect(connectionInfo);
+    public async connect(
+        resource: Resource,
+        kernelName?: string,
+        cancelToken?: CancellationToken
+    ): Promise<IJupyterKernelSpec | undefined> {
+        try {
+            // Try to start up our raw session, allow for cancellation or timeout
+            // Notebook Provider level will handle the thrown error
+            const rawSessionStart = await waitForPromise(
+                Promise.race([
+                    this.startRawSession(resource, kernelName),
+                    createPromiseFromCancellation({
+                        cancelAction: 'reject',
+                        defaultValue: new CancellationError(),
+                        token: cancelToken
+                    })
+                ]),
+                5_000
+            );
+
+            // Only connect our session if we didn't cancel or timeout
+            if (rawSessionStart instanceof CancellationError) {
+                traceInfo('Starting of raw session cancelled by user');
+                throw rawSessionStart;
+            } else if (rawSessionStart === null) {
+                traceError('Raw session failed to start in given timeout');
+                throw new Error(localize.DataScience.sessionDisposed());
+            } else {
+                traceInfo('Raw session started and connected');
+                this.currentSession = rawSessionStart;
+                this.session = this.currentSession.session;
+            }
+        } catch {
+            traceError('Failed to connect raw kernel session');
```

Comment:
```suggestion
        } catch (ex) {
            traceError('Failed to connect raw kernel session', ex);
```

I don't think we should be swallowing the original exception.

Reply:
I was doing this on purpose initially as as was trying to narrow down the possible exception cases to a more understandable / localized message (our sessionDisposed message that we use fairly often to indicate a disconnect)but yeah, especially with kernel launcher involved I shouldn't be doing that. I'll rethrow here and flesh out the kernel launcher exceptions some.

Reply:
Some throws in here probably will still use the SessionDisconnect message, but they will do that explicitly, not by swallowing.